python/cufinufft/cufinufft/_plan.py

```diff
@@ -0,0 +1,373 @@
+        ier = self._setpts(M, *fpts_axes[:3], 0, None, None, None, self._plan)
+
+        self.nj = M
+
+        if ier != 0:
+            raise RuntimeError('Error setting non-uniform points.')
+
+    def execute(self, data, out=None):
+        """
+        Execute the plan
+
+        Performs the NUFFT specified at plan instantiation with the points set
+        by ``setpts``. For type-1 and type-3 transforms, the input is a set of
+        source strengths, while for a type-2 transform, it consists of an
+        array of size ``n_modes``. If ``n_trans`` is greater than one,
+        ``n_trans`` inputs are expected, stacked along the first axis.
+
+        Args:
+            data    (complex[M], complex[n_transf, M], complex[n_modes], or complex[n_transf, n_modes]): The input source strengths
+                    (type 1) or source modes (type 2).
+            out     (complex[n_modes], complex[n_transf, n_modes], complex[M], or complex[n_transf, M], optional): The array where the
+                    output is stored. Must be of the right size.
+
+        Returns:
+            complex[n_modes], complex[n_transf, n_modes], complex[M], or complex[n_transf, M]: The output array of the transform(s).
+        """
+
+        _data = _ensure_array_type(data, "data", self.dtype)
+        _out = _ensure_array_type(out, "out", self.dtype, output=True)
+
+        if self.type == 1:
+            req_data_shape = (self.n_trans, self.nj)
+            req_out_shape = self.n_modes
+        elif self.type == 2:
+            req_data_shape = (self.n_trans, *self.n_modes)
+            req_out_shape = (self.nj,)
+
+        _data, data_shape = _ensure_array_shape(_data, "data", req_data_shape,
+                                                allow_reshape=True)
+        if self.type == 1:
+            batch_shape = data_shape[:-1]
+        else:
+            batch_shape = data_shape[:-self.dim]
+
+        req_out_shape = batch_shape + req_out_shape
+
+        if out is None:
+            _out = GPUArray(req_out_shape, dtype=self.dtype)
+        else:
+            _out = _ensure_array_shape(_out, "out", req_out_shape)
+
+        if self.type == 1:
+            ier = self._exec_plan(data.ptr, _out.ptr, self._plan)
+        elif self.type == 2:
+            ier = self._exec_plan(_out.ptr, data.ptr, self._plan)
+
+        if ier != 0:
+            raise RuntimeError('Error executing plan.')
+
+        return _out
+
+    def __del__(self):
+        """
+        Destroy this instance's associated plan and storage.
+        """
+
+        # If the process is exiting or we've already cleaned up plan, return.
+        if exiting or self._plan is None:
+            return
+
+        ier = self._destroy_plan(self._plan)
+
+        if ier != 0:
+            raise RuntimeError('Error destroying plan.')
+
+        # Reset plan to avoid double destroy.
+        self._plan = None
+
+        # Reset our reference.
+        self.references = []
```

Comment:
this should be `self._references`, though anything but `_destroy_plan` should be pedantic anyway.

Reply:
Right. Will update.